File: internal/scraper/jurisdictions/courtlistener.go
package jurisdictions

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/gongahkia/kite/internal/scraper"
	"github.com/gongahkia/kite/pkg/errors"
	"github.com/gongahkia/kite/pkg/models"
)

// CourtListenerScraper implements scraping for CourtListener (US Federal & State Courts)
type CourtListenerScraper struct {
	*scraper.BaseScraper
	baseURL string
	client  *http.Client
}

// NewCourtListenerScraper creates a new CourtListener scraper
func NewCourtListenerScraper() *CourtListenerScraper {
	baseURL := "https://www.courtlistener.com"
	base := scraper.NewBaseScraper(
		"CourtListener",
		"United States",
		baseURL,
		20, // 20 requests per minute
	)

	return &CourtListenerScraper{
		BaseScraper: base,
		baseURL:     baseURL,
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// SearchCases searches for cases matching the query
func (cls *CourtListenerScraper) SearchCases(ctx context.Context, query scraper.SearchQuery) ([]*models.Case, error) {
	// Build search URL
	searchURL, err := cls.buildSearchURL(query)
	if err != nil {
		return nil, errors.ParsingError("failed to build search URL", err)
	}

	// Check robots.txt
	allowed, err := cls.BaseScraper.client.CheckRobots(ctx, "/")
	if err != nil || !allowed {
		return nil, errors.ErrRobotsDisallowed
	}

	// Wait for rate limit
	if err := cls.BaseScraper.client.rateLimiter.Wait(ctx); err != nil {
		return nil, errors.RateLimitError("rate limit exceeded")
	}

	// Make HTTP request
	req, err := http.NewRequestWithContext(ctx, "GET", searchURL, nil)
	if err != nil {
		return nil, errors.NetworkError("failed to create request", err)
	}

	req.Header.Set("User-Agent", cls.BaseScraper.client.userAgent)

	resp, err := cls.client.Do(req)
	if err != nil {
		return nil, errors.NetworkError("failed to fetch search results", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, errors.NetworkError(fmt.Sprintf("unexpected status code: %d", resp.StatusCode), nil)
	}

	// Parse HTML
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, errors.ParsingError("failed to parse HTML", err)
	}

	// Extract cases from search results
	cases := make([]*models.Case, 0)

	doc.Find("article.search-document").Each(func(i int, s *goquery.Selection) {
		if query.Limit > 0 && len(cases) >= query.Limit {
			return
		}

		caseData := cls.extractCaseFromSearchResult(s)
		if caseData != nil {
			cases = append(cases, caseData)
		}
	})

	return cases, nil
}

// GetCaseByID retrieves a specific case by its ID
func (cls *CourtListenerScraper) GetCaseByID(ctx context.Context, caseID string) (*models.Case, error) {
	// CourtListener uses opinion IDs in URLs
	caseURL := fmt.Sprintf("%s/opinion/%s/", cls.baseURL, caseID)

	// Check robots.txt
	allowed, err := cls.BaseScraper.client.CheckRobots(ctx, "/opinion/")
	if err != nil || !allowed {
		return nil, errors.ErrRobotsDisallowed
	}

	// Wait for rate limit
	if err := cls.BaseScraper.client.rateLimiter.Wait(ctx); err != nil {
		return nil, errors.RateLimitError("rate limit exceeded")
	}

	// Make HTTP request
	req, err := http.NewRequestWithContext(ctx, "GET", caseURL, nil)
	if err != nil {
		return nil, errors.NetworkError("failed to create request", err)
	}

	req.Header.Set("User-Agent", cls.BaseScraper.client.userAgent)

	resp, err := cls.client.Do(req)
	if err != nil {
		return nil, errors.NetworkError("failed to fetch case", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return nil, errors.ErrNotFound
	}

	if resp.StatusCode != http.StatusOK {
		return nil, errors.NetworkError(fmt.Sprintf("unexpected status code: %d", resp.StatusCode), nil)
	}

	// Parse HTML
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, errors.ParsingError("failed to parse HTML", err)
	}

	// Extract case details
	return cls.extractCaseDetails(doc, caseID, caseURL)
}

// GetCasesByDateRange retrieves cases within a date range
func (cls *CourtListenerScraper) GetCasesByDateRange(ctx context.Context, startDate, endDate time.Time, limit int) ([]*models.Case, error) {
	query := scraper.SearchQuery{
		StartDate: &startDate,
		EndDate:   &endDate,
		Limit:     limit,
	}

	return cls.SearchCases(ctx, query)
}

// IsAvailable checks if CourtListener is available
func (cls *CourtListenerScraper) IsAvailable(ctx context.Context) bool {
	req, err := http.NewRequestWithContext(ctx, "HEAD", cls.baseURL, nil)
	if err != nil {
		return false
	}

	resp, err := cls.client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()

	return resp.StatusCode == http.StatusOK
}

// buildSearchURL builds the search URL with query parameters
func (cls *CourtListenerScraper) buildSearchURL(query scraper.SearchQuery) (string, error) {
	params := url.Values{}

	if query.Query != "" {
		params.Set("q", query.Query)
	}

	if query.Court != "" {
		params.Set("court", query.Court)
	}

	if query.StartDate != nil {
		params.Set("filed_after", query.StartDate.Format("2006-01-02"))
	}

	if query.EndDate != nil {
		params.Set("filed_before", query.EndDate.Format("2006-01-02"))
	}

	params.Set("type", "o") // o = opinions
	params.Set("order_by", "score desc")

	searchURL := fmt.Sprintf("%s/?%s", cls.baseURL, params.Encode())
	return searchURL, nil
}

// extractCaseFromSearchResult extracts case data from a search result item
func (cls *CourtListenerScraper) extractCaseFromSearchResult(s *goquery.Selection) *models.Case {
	c := models.NewCase()

	// Extract case name
	caseName := s.Find("h3.bottom a").Text()
	c.CaseName = strings.TrimSpace(caseName)

	// Extract case URL and ID
	caseURL, exists := s.Find("h3.bottom a").Attr("href")
	if exists {
		c.URL = cls.baseURL + caseURL
		// Extract ID from URL (e.g., /opinion/123456/case-name/)
		parts := strings.Split(strings.Trim(caseURL, "/"), "/")
		if len(parts) >= 2 {
			c.ID = parts[1]
		}
	}

	// Extract court
	court := s.Find(".meta-data-header").Text()
	c.Court = strings.TrimSpace(court)

	// Extract date
	dateStr := s.Find(".meta-data-header time").AttrOr("datetime", "")
	if dateStr != "" {
		if date, err := time.Parse("2006-01-02", dateStr); err == nil {
			c.DecisionDate = &date
		}
	}

	// Extract snippet/summary
	snippet := s.Find(".snippet").Text()
	c.Summary = strings.TrimSpace(snippet)

	// Set metadata
	c.Jurisdiction = "United States"
	c.SourceDatabase = "CourtListener"
	c.ScrapedAt = time.Now()
	c.LastUpdated = time.Now()
	c.Language = "en"
	c.Status = models.CaseStatusActive

	return c
}

// extractCaseDetails extracts detailed case information from a case page
func (cls *CourtListenerScraper) extractCaseDetails(doc *goquery.Document, caseID, caseURL string) (*models.Case, error) {
	c := models.NewCase()

	c.ID = caseID
	c.URL = caseURL

	// Extract case name
	caseName := doc.Find("h1.text-center").Text()
	c.CaseName = strings.TrimSpace(caseName)

	// Extract court
	court := doc.Find(".meta-data-header a").First().Text()
	c.Court = strings.TrimSpace(court)

	// Extract date
	dateStr := doc.Find("time").AttrOr("datetime", "")
	if dateStr != "" {
		if date, err := time.Parse("2006-01-02", dateStr); err == nil {
			c.DecisionDate = &date
		}
	}

	// Extract judges
	doc.Find(".author a").Each(func(i int, s *goquery.Selection) {
		judge := strings.TrimSpace(s.Text())
		if judge != "" {
			c.Judges = append(c.Judges, judge)
		}
	})

	// Extract docket number
	docket := doc.Find(".docket-number").Text()
	c.Docket = strings.TrimSpace(docket)

	// Extract case text
	fullText := doc.Find("#opinion-content").Text()
	c.FullText = strings.TrimSpace(fullText)

	// Set metadata
	c.Jurisdiction = "United States"
	c.SourceDatabase = "CourtListener"
	c.ScrapedAt = time.Now()
	c.LastUpdated = time.Now()
	c.Language = "en"
	c.Status = models.CaseStatusActive

	return c, nil
}

File: archive/v3-nim/src/kite/scrapers/courtlistener.nim
import std/[json, times, strutils]
import chronicles
from ../../tja/base_scraper import BaseScraper, initBaseScraper, makeRequest
from ../../tja/data_models import CaseData, initCaseData
from ../../tja/helpers import sanitizeText


type
  CourtListenerScraper* = ref object of BaseScraper

method baseUrl*(self: CourtListenerScraper): string = "https://www.courtlistener.com"
method jurisdiction*(self: CourtListenerScraper): string = "United States"

proc newCourtListenerScraper*(): CourtListenerScraper =
  result = CourtListenerScraper(initBaseScraper())

proc parseSearchItem(item: JsonNode; baseUrl: string): CaseData =
  var c = initCaseData()
  c.caseName = sanitizeText(item.getOrDefault("caseName").getStr)
  c.caseId = $item.getOrDefault("id").getInt
  if item.hasKey("dateFiled"):
    try:
      c.date = parse("yyyy-MM-dd", item["dateFiled"].getStr)
    except: discard
  c.court = sanitizeText(item.getOrDefault("court").getStr)
  if item.hasKey("absolute_url"):
    c.url = baseUrl & item["absolute_url"].getStr
  if item.hasKey("citation"):
    c.citations = @[item["citation"].getStr]
  if item.hasKey("judge"):
    c.judges = @[sanitizeText(item["judge"].getStr)]
  c.summary = sanitizeText(item.getOrDefault("snippet").getStr)
  c.jurisdiction = "United States"
  c.caseType = sanitizeText(item.getOrDefault("status").getStr)
  return c

proc searchCases*(self: CourtListenerScraper; query: string; startDate = ""; endDate = ""; court = ""; limit = 100): seq[CaseData] =
  var params: seq[(string,string)] = @[("format","json"), ("order_by","-date_filed")]
  if query.len > 0: params.add ("q", query)
  if startDate.len > 0: params.add ("filed_after", startDate)
  if endDate.len > 0: params.add ("filed_before", endDate)
  if court.len > 0: params.add ("court", court)
  if limit > 0: params.add ("count", $(min(limit, 1000)))
  let url = self.baseUrl() & "/api/rest/v3/search/"
  let resp = self.makeRequest(url, params=params)
  let data = parseJson(resp.body)
  let results = data.getOrDefault("results")
  if results.kind == JArray:
    for item in results.items:
      try:
        result.add parseSearchItem(item, self.baseUrl())
      except: continue
  return result

proc getCaseById*(self: CourtListenerScraper; caseId: string): CaseData =
  if caseId.len == 0: return initCaseData()
  var url = self.baseUrl() & "/api/rest/v3/opinions/" & caseId & "/"
  try:
    let resp = self.makeRequest(url, params = @[("format","json")])
    let data = parseJson(resp.body)
    # cluster lookup
    if data.hasKey("cluster"):
      let clusterUrl = data["cluster"].getStr
      let cresp = self.makeRequest(clusterUrl, params = @[("format","json")])
      let cdata = parseJson(cresp.body)
      var c = initCaseData()
      c.caseName = sanitizeText(cdata.getOrDefault("case_name").getStr)
      if cdata.hasKey("date_filed"):
        try: c.date = parse("yyyy-MM-dd", cdata["date_filed"].getStr) except: discard
      c.court = sanitizeText(cdata.getOrDefault("court").getStr)
      c.url = self.baseUrl() & cdata.getOrDefault("absolute_url").getStr
      c.caseId = $data.getOrDefault("id").getInt
      c.jurisdiction = self.jurisdiction()
      return c
  except: discard
  # try cluster endpoint directly
  url = self.baseUrl() & "/api/rest/v3/clusters/" & caseId & "/"
  try:
    let resp = self.makeRequest(url, params = @[("format","json")])
    let data = parseJson(resp.body)
    var c = initCaseData()
    c.caseName = sanitizeText(data.getOrDefault("case_name").getStr)
    if data.hasKey("date_filed"):
      try: c.date = parse("yyyy-MM-dd", data["date_filed"].getStr) except: discard
    c.court = sanitizeText(data.getOrDefault("court").getStr)
    c.url = self.baseUrl() & data.getOrDefault("absolute_url").getStr
    c.caseId = $data.getOrDefault("id").getInt
    c.jurisdiction = self.jurisdiction()
    return c
  except: discard
  return initCaseData()
